src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java

```diff
@@ -386,21 +380,18 @@ public List<Course> getAllCoursesForManagementOverview(@RequestParam(defaultValu
      */
     @GetMapping("courses/for-registration")
     @PreAuthorize("hasRole('USER')")
-    public List<Course> getAllCoursesToRegister() {
-        log.debug("REST request to get all currently active Courses that are not online courses");
+    public List<Course> getAllCoursesForRegistration() {
+        log.debug("REST request to get all currently active courses that are not online courses");
         User user = userRepository.getUserWithGroupsAndAuthoritiesAndOrganizations();
 
-        List<Course> allRegisteredCourses = courseService.findAllActiveForUser(user);
-        List<Course> allCoursesToRegister = courseRepository.findAllCurrentlyActiveNotOnlineAndRegistrationEnabledWithOrganizationsAndPrerequisites();
-        return allCoursesToRegister.stream().filter(course -> {
-            // further, check if the course has been assigned to any organization and if yes,
-            // check if user is member of at least one of them
-            if (course.getOrganizations() != null && !course.getOrganizations().isEmpty()) {
-                return courseRepository.checkIfUserIsMemberOfCourseOrganizations(user, course);
-            }
-            else {
-                return true;
-            }
+        Set<Course> allRegisteredCourses = courseService.findAllActiveForUser(user);
+        List<Course> allCoursesToPotentiallyRegister = courseRepository.findAllActiveNotOnlineAndRegistrationEnabledWithOrganizationsAndPrerequisites();
+        // check whether registration is actually possible for each of the courses
+        return allCoursesToPotentiallyRegister.stream().filter(course -> {
+            boolean selfRegistrationAllowed = AuthorizationCheckService.RegistrationAuthorization.ALLOWED
+                    .equals(authCheckService.getUserRegistrationAuthorizationForCourse(user, course));
+            boolean isAlreadyInCourse = allRegisteredCourses.contains(course);
+            return selfRegistrationAllowed && !isAlreadyInCourse;
         }).filter(course -> !allRegisteredCourses.contains(course)).toList();
```

Comment:
```suggestion
        }).toList();
```
I don't think you need the second filter now, since you are already filtering by !isAlreadyInCourse in the first one.